### ISSUE TITLE
Ability to switch networks from a `NetworkId`

### ISSUE THREAD

**Issue by trechriron** [near/wallet-selector#928]
### Description

Is your feature request related to a problem? Please describe. Currently, the network can be overridden through `selector.options.network = getNetworkPreset(network);` (UT if that's functional or if any missing side effects) but `getNetworkPreset` is not exposed from the library.

Describe the solution you'd like Would be nice to have a helper function to update the network of the wallet selector using a `NetworkId` to avoid having to copy code (and assume no side effects of modifying options).

Might be worthwhile to consider #403 with this, as the changes are somewhat connected (although this affects the selector type where the other is relating to the modal).

Describe alternatives you've considered

This could possibly just be done by replacing the wallet selector instance, but unsure if there are unintended side effects from doing this. If this is the intended flow, feel free to close this issue 

Acceptance criteria Network can be changed through a `NetworkId` once the wallet selector is already initialized

Amir Saran 
February 2, 2023 at 3:01 AM
Edited

exalate-issue-sync[bot] commented: 

Amir Saran commented:

Regarding [Daryl Collins](https://pagodaplatform.atlassian.net/secure/ViewProfile.jspa?accountId=6139108a24ba8b00701167be) comment:

The initial implementation of network switching was using multiple wallet-selector instances. After talking with the Pagoda team we came to the conclusion that we want to have the network switching done inside the wallet-selector modal.

Having it inside the wallet-selector modal means that we need to add a `networkChanged` event so that the dApp can update its view after network switching. This is not an ideal and intuitive solution since the developer needs to add event listeners to all wallet-selector instances with same functionality.

Daryl Collins 
January 17, 2023 at 10:10 AM
Edited

exalate-issue-sync[bot] commented: 

Daryl Collins commented:

Is there any problem with Austin’s proposed solution to this?

This could possibly just be done by replacing the wallet selector instance, but unsure if there are unintended side effects from doing this. If this is the intended flow, feel free to close this issue 

Some of the reasons that this is currently driven  from the dapp scope  are:

Note that it is unlikely-to-impossible for the exact same account to exist on both testnet and mainnet as testnet seed phrases hash differently than mainnet ones do for deriving implicit account names

There is a different top-level-account that is responsible for named account creation on testnet (.testnet) than on mainnet (.near).

Also, most dapps are deployed in such a way that an instance of the dapp is always tied to a specific network ID, with an assumption that they are running in one  specific  network. Those apps will have no awareness of how to react to a network change, and would be broken if the wallet used accounts from a different chain than they are built to communicate with.

Amir Saran 
December 14, 2022 at 4:30 AM
Edited

exalate-issue-sync[bot] commented: 

Amir Saran commented:

Currently this issue is blocked by the injected wallet implementation. There is no way to tell the injected wallets that the dApp has changed networks. This will cause a synching problem since the wallet and the dApp are using different active networks. In the case of Sender wallet if the user tries to connect on testnet network but the wallet network is set to mainnet it will not work.

This can be done with these two options:

Provide a network parameter when signing in which will tell the wallet what network the user intends to use.

Injected wallets create a function for changing the network from the dApp side. 

These options will have checks because the user may not have a connected account with a particular network in the wallet.

These changes should also be done in the Injected Wallet NEP.

Exalate 
October 31, 2022 at 3:18 AM

kujtimprenkuSQA commented: 

> +1 - the ability to swap between testnet/mainnet as needed is pretty key within our app.

@jessupjn I would like to know how did you handle switching networks in your app before using Wallet Selector?

> Then the options can be updated via selector.options.network = newNetwork and picked up by the wallets when needed

As mentioned above at the moment there is no way to "properly" switch the network in Wallet Selector.

Just overriding the network `selector.options.network` is not enough because there's no event being emitted by doing so.

The [`provider`](https://github.com/near/wallet-selector/blob/main/packages/core/src/lib/wallet-selector.ts#L27) service we pass to the modules would not be updated this way.

The wallet connection when you first signed in would not be updated.

The network on the wallet will not be updated to the one you're changing in the wallet selector.

Exalate 
October 27, 2022 at 6:28 AM

jessupjn commented: 

Then the options can be updated via `selector.options.network = newNetwork` and picked up by the wallets when needed
